### PR TITLE
Interface wrappers + State APIs (and two new samples).

### PR DESCRIPTION
**Pull request by JeroMiya**
This implements the following issues:
https://github.com/OSVR/Managed-OSVR/issues/7
https://github.com/OSVR/Managed-OSVR/issues/1
https://github.com/OSVR/OSVR-Unity/issues/5 (duplicate)

Done together because the state APIs are included in the wrappers. Also includes a utility class InterfaceAdapter, to be used by OSVR-MonoGame and OSVR-Unity to more easily adapt the raw report/state types to MonoGame/Unity types, respectively. This should greatly reduce boilerplate code in both frameworks. Finally, two new samples: one for the state apis and one for InterfaceAdapter.

### CODE REVIEW

**Comment by rpavlik** [comment]
So this looks really neat, thanks!  A couple of notes:
- There is actually built-in support for state access, you don't need to store it yourself, just wrap the functions: http://resource.osvr.com/docs/OSVR-Core/InterfaceStateC_8h.html  That will avoid duplication of code, lets us handle more complex state storage entirely inside the core, and avoids native-to-managed callbacks just to update the state storage.
- Does C# have anything like C++ template metaprogramming or traits classes? I ask because in the C++ core code, we have a header that basically sets up traits classes that, given a report type, compute the corresponding state type. Might make the interface adapter usage safer.

**Comment by JeroMiya** [comment]
Actually the native state API P/Invoke mappings are included in this PR. InterfaceBase has a constructor that takes a reference to the native state getter function as a delegate, which it uses in the base class implementation of GetState(). All of the Interface wrappers I added in this PR are passing in the relevant native state API to use in their constructors.

There might be an appropriate motivation for caching the state in managed code, which would be for performance reasons (to avoid the native/managed marshalling overhead), but I figured we'd wait until we actually measured a performance problem before we start fixing it. The comment on IInterface.GetState about re-using the return value of the GetState function is just a recommendation - it's the same one made by the XNA teams about the various input APIs like Keyboard.GetState()/etc... and it's just recommending that you call it at most once per Update cycle (or whatever the equivalent is in your framework of choice) and re-use the value instead of making multiple calls to GetState, if possible. The GetState method itself calls the native state API every time.

C# doesn't yet have metaprogramming capabilities. In the .Net world, so far only F# has this with type providers but F# has an additional library dependency that I wanted to avoid. Traditionally this has been done by teams with T4 templates, which Visual Studio supports for code generation, but this isn't cross platform. The cross-platform alternative is to just have a script that generates the C# code as you have in the OSVR-Unity repository. Barring that, I made an effort to reduce the amount of boilerplate in InterfaceBase derived classes, so it doesn't pass my personal "boilerplate vs code-gen" threshold. However I could look into it if you like.

Question for OSVR guys: What is the "sensor" field of the report types? Are you expecting client code to need it? If so, are you planning on adding it to the native state APIs?

If client code doesn't need it, then I could remove the separate TReport and TState types in the generic IInterface type, which would be much simpler for users of the API. For instance, instead of PositionInterface being an IInterface<PositionReport, Vec3>, it would just be an IInterface<Vec3> and both callbacks and the state APIs would both report/return Vec3s. Or, alternatively, we could for example modify the state APIs in the native ClientKit to return a PositionReport instead of just the Vec3 for the position.

**Comment by JeroMiya** [comment]
Correction: apparently MonoDevelop now supports T4 templates natively! So, this might be a good solution to generate the InterfaceBase sub-classes for each interface type. T4 templates are built along with the project, so they are a convenient code-generation tool.

**Comment by rpavlik** [comment]
The sensor field is specifically only in Report types and not State types. The idea is that you could register a callback on an entire interface (all buttons from a device, for example), and so you wouldn't necessarily know what you're getting from when you get called. (This is not a common use case, more common in analysis plugins than client apps certainly, for filtering, etc, but supported.) On the other hand, if you're requesting state, you can only do that from a single sensor, so you know what you're getting.

Thought I saw some place where you were storing state, not sure why/where, will look through again, that's why I assumed you were using callbacks and storing state locally.

Our current presumption that the state interface will be better performing for some uses is because the sample rate of many trackers is much higher (250Hz+) than framerate, so each frame ends up getting multiple callbacks for a single sensor when it only ends up caring about the most recent one - thus more native->managed callbacks. In any case, regardless of performance differences, using callbacks vs. state is a useful option to offer a developer since their app model may favor one over another.

**Comment by JeroMiya** [comment]
Thank you - that makes sense. Is there a way to get the Int32 sensor value for a given interface, or convert that value to the OSVR path?

Either way, if all of the report types have this same sensor value, maybe I can adjust the signatures of the generic delegates to separate out the sensor value from the state value in the event handler delegate so that I can eliminate the extra report type param. I'll look into this tonight - I may have one more checkin for this pull request, so please hold off on merging it.

**Comment by rpavlik** [comment]
Not at the moment - and there may be multiple aliases (paths) pointing to a single sensor, so the result would not necc. be unique. Have considered adding it as a callback parameter, but that wouldn't really solve the issue of subscribing to a whole interface.

**Comment by JeroMiya** [comment]
OK, I've added the sensor argument to the event delegate type and reduced type params on IInterface and InterfaceBase from 2 (TReport and TState) to 1 (T, previously TState). This pull request is again ready for evaluation/merge. Thanks!

**Comment by JeroMiya** [comment]
Is there any other feedback on this pull request? I would like to get started on updating OSVR-MonoGame and OSVR-Unity for these changes.

File: ExampleClients/ButtonState/ButtonState.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+/// </copyright>
+
+﻿using System;
+using OSVR.ClientKit;
+
+namespace ButtonState
+{
+    class MainClass
+    {
+        const double Speed = 2.0;
```
Comment (rpavlik) [inline ExampleClients/ButtonState/ButtonState.cs]: Is this getting used?

**Comment by rpavlik** [comment]
Sorry I didn't get back to you sooner on this. Generally looks really good, just a few notes/questions:
- Copyright headers - can at least be 2014, can be your name up there.
- example app client id's: should be `com.osvr.whatever...`
- Is there a reason that we need `Start` methods on the interface objects? They aren't on the underlying objects.
- Any way to reduce the amount of boilerplate-like code needed for each interface? (This might be a post-merge thing)  Something like a `GetInterface<PositionInterface>()` generic call maybe?

**Comment by JeroMiya** [comment]
I considered not having start methods. On the managed side, the two alternatives are:
- Immediately register a native callback in the interface constructor (i.e. move Start code into constructor).
- Implement event subscribe/unsubscribe blocks, such that the native callback is registered the first time a user subscribes to the managed event.

The first option may have performance implications if the user is only interested in the state APIs, and I wasn't sure if there would be timing issues with dependency injection libraries (like the one in Unity). The second option is ideal from a usability standpoint, but increases internal complexity.

I chose a conservative route and required Start to be called explicitly, however the custom event subscribe/unsubscribe blocks would be easier to use. I would make Start a protected abstract method, used by a common event implementation.

I'll look into generic GetInterface method idea. I might be able to do it without reflection using type constraints. Reflection is a little tricky in existing (.Net native, Mono, CSNative) and upcoming (LLILC) AOT engines, and is best avoided if there are alternatives.

**Comment by rpavlik** [comment]
I think I used the second option for what I did in Unity - ended up being basically an extra line that if the native callback was null, register it, or something like that. No need to un-register the callback - because first of all, you can't (you just free the interface object), and second of all, you shouldn't need to (getting rid of the interface object takes care of it for you)

**Comment by JeroMiya** [comment]
How about "Copyright 2014, 2015 Sensics, Inc. and contributors"?

**Comment by JeroMiya** [comment]
Assuming the copyright change is ok, I think these two commits should address your feedback, aside from the generic `GetInterface<T>` method you proposed.

**Comment by rpavlik** [comment]
Ah, cool, didn't know you could do that with events like properties. Looks good, merging! (Does look like some formatting changes made it in, but not going to nitpick them because I don't know of a scriptable tool to enforce/apply it.)

**Comment by JeroMiya** [comment]
I looked into a generic GetInterface<T> implementation. I can't implement it using generic type constraints - only through reflection. I'd like to avoid reflection because it places an additional burden if the user is running .Net through one of the AOT runtimes, which all require you to declare ahead of time what reflection metadata you'll need. Instead, I'd like to keep using the static method or extension method route. It's a little more boilerplate, but we might be able to mitigate that with a T4 template.

For reference, MSDN docs on reflection and .Net Native: https://msdn.microsoft.com/en-us/library/dn600640(v=vs.110).aspx

If we did go the reflection route, it would look something like this (possibly with some additional sanity checks, perhaps behind an `#if DEBUG` block):

``` C#
            public T GetInterfaceType<T>(string path)
            {
                var type = typeof(T);
                var ifaceCtor = type.GetConstructor(new Type[] { typeof(Interface) });

                if(ifaceCtor == null)
                {
                    throw new InvalidOperationException(
                        String.Format("Type {0} does not have a public constructor taking a single Interface argument.", type.Name));
                }

                var rawInterface = this.getInterface(path);
                var wrappedInterface = (T)ifaceCtor.Invoke(new object[] { rawInterface });
                return wrappedInterface;
            }
```

**Comment by rpavlik** [comment]
Oh ick, you're right, that is unpleasant.  Apparently I'm spoiled by C++ templates vs .net generics.

**Pull request merged**